File: Source/AlphaTab/Audio/Synth/Synthesis/Envelope.cs
// The SoundFont loading and Audio Synthesis is based on TinySoundFont, licensed under MIT,
// developed by Bernhard Schelling (https://github.com/schellingb/TinySoundFont)

// C# port for alphaTab: (C) 2019 by Daniel Kuschny
// Licensed under: MPL-2.0

/*
 * LICENSE (MIT)
 *
 * Copyright (C) 2017, 2018 Bernhard Schelling
 * Based on SFZero, Copyright (C) 2012 Steve Folta (https://github.com/stevefolta/SFZero)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify, merge,
 * publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons
 * to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
 * USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

using AlphaTab.Audio.Synth.Util;

namespace AlphaTab.Audio.Synth.Synthesis
{
    internal class Envelope
    {
        public float Delay { get; set; }
        public float Attack { get; set; }
        public float Hold { get; set; }
        public float Decay { get; set; }
        public float Sustain { get; set; }
        public float Release { get; set; }
        public float KeynumToHold { get; set; }
        public float KeynumToDecay { get; set; }

        public Envelope()
        {
        }

        public Envelope(Envelope other)
        {
            Delay = other.Delay;
            Attack = other.Attack;
            Hold = other.Hold;
            Decay = other.Decay;
            Sustain = other.Sustain;
            Release = other.Release;
            KeynumToHold = other.KeynumToHold;
            KeynumToDecay = other.KeynumToDecay;
        }

        public void Clear()
        {
            Delay = 0;
            Attack = 0;
            Hold = 0;
            Decay = 0;
            Sustain = 0;
            Release = 0;
            KeynumToHold = 0;
            KeynumToDecay = 0;
        }

        public void EnvToSecs(bool sustainIsGain)
        {
            // EG times need to be converted from timecents to seconds.
            // Pin very short EG segments.  Timecents don't get to zero, and our EG is
            // happier with zero values.
            Delay = (Delay < -11950.0f ? 0.0f : SynthHelper.Timecents2Secs(Delay));
            Attack = (Attack < -11950.0f ? 0.0f : SynthHelper.Timecents2Secs(Attack));
            Release = (Release < -11950.0f ? 0.0f : SynthHelper.Timecents2Secs(Release));

            // If we have dynamic hold or decay times depending on key number we need
            // to keep the values in timecents so we can calculate it during startNote
            if (KeynumToHold == 0)
            {
                Hold = (Hold < -11950.0f ? 0.0f : SynthHelper.Timecents2Secs(Hold));
            }

            if (KeynumToDecay == 0)
            {
                Decay = (Decay < -11950.0f ? 0.0f : SynthHelper.Timecents2Secs(Decay));
            }

            if (Sustain < 0.0f)
            {
                Sustain = 0.0f;
            }
            else if (sustainIsGain)
            {
                Sustain = SynthHelper.DecibelsToGain(-Sustain / 10.0f);
            }
            else
            {
                Sustain = 1.0f - (Sustain / 1000.0f);
            }
        }
    }
}

File: src/synth/synthesis/Envelope.ts
// The SoundFont loading and Audio Synthesis is based on TinySoundFont, licensed under MIT,
// developed by Bernhard Schelling (https://github.com/schellingb/TinySoundFont)
// TypeScript port for alphaTab: (C) 2020 by Daniel Kuschny
// Licensed under: MPL-2.0
import { SynthHelper } from '@src/synth/SynthHelper';

export class Envelope {
    public delay: number = 0;
    public attack: number = 0;
    public hold: number = 0;
    public decay: number = 0;
    public sustain: number = 0;
    public release: number = 0;
    public keynumToHold: number = 0;
    public keynumToDecay: number = 0;

    public constructor(other?: Envelope) {
        if (other) {
            this.delay = other.delay;
            this.attack = other.attack;
            this.hold = other.hold;
            this.decay = other.decay;
            this.sustain = other.sustain;
            this.release = other.release;
            this.keynumToHold = other.keynumToHold;
            this.keynumToDecay = other.keynumToDecay;
        }
    }

    public clear(): void {
        this.delay = 0;
        this.attack = 0;
        this.hold = 0;
        this.decay = 0;
        this.sustain = 0;
        this.release = 0;
        this.keynumToHold = 0;
        this.keynumToDecay = 0;
    }

    public envToSecs(sustainIsGain: boolean): void {
        // EG times need to be converted from timecents to seconds.
        // Pin very short EG segments.  Timecents don't get to zero, and our EG is
        // happier with zero values.
        this.delay = this.delay < -11950.0 ? 0.0 : SynthHelper.timecents2Secs(this.delay);
        this.attack = this.attack < -11950.0 ? 0.0 : SynthHelper.timecents2Secs(this.attack);
        this.release = this.release < -11950.0 ? 0.0 : SynthHelper.timecents2Secs(this.release);

        // If we have dynamic hold or decay times depending on key number we need
        // to keep the values in timecents so we can calculate it during startNote
        if (this.keynumToHold === 0) {
            this.hold = this.hold < -11950.0 ? 0.0 : SynthHelper.timecents2Secs(this.hold);
        }

        if (this.keynumToDecay === 0) {
            this.decay = this.decay < -11950.0 ? 0.0 : SynthHelper.timecents2Secs(this.decay);
        }
        
        if (this.sustain < 0.0) {
            this.sustain = 0.0;
        } else if (sustainIsGain) {
            this.sustain = SynthHelper.decibelsToGain(-this.sustain / 10.0);
        } else {
            this.sustain = 1.0 - this.sustain / 1000.0;
        }
    }
}
